SSA/Projects/CIRCT/comb/comb.lean

```diff
@@ -0,0 +1,539 @@
+import SSA.Core.MLIRSyntax.EDSL
+import SSA.Projects.CIRCT.Stream.Stream
+import SSA.Projects.CIRCT.Stream.WeakBisim
+
+
+namespace Comb
+
+
+def add {w : Nat} (l : List (BitVec w)) : BitVec w :=
+  List.foldr BitVec.add (0#w) l
+
+def and {w : Nat} (l : List (BitVec w)) : BitVec w :=
+  List.foldr BitVec.and (0#w) l
+
+/-- Concatenate a list of bitvecs `xs`, where the length of bitvec xs[i] is given by
+  element ls[i] in a list of nat `ls`-/
+def concat {ls : List Nat} (xs : HVector BitVec ls) : BitVec (List.sum ls) :=
+  match (xs) with
+  | (.nil) => 0#0
+  | (.cons x xs) =>
+     let xsSum := concat xs
+     x ++ xsSum
+
+def divs (x y : BitVec w) : BitVec w :=
+  BitVec.sdiv x y
+
+def divu (x y : BitVec w) : BitVec w :=
+  BitVec.udiv x y
+
+def extract (x : BitVec w) (lb : Nat) : BitVec (w - lb) :=
+  BitVec.truncate (w - lb) (BitVec.ushiftRight x lb)
+
+def icmp (x y : BitVec w) (op : Nat) : Bool :=
+  match op with
+  | 0 => x == y
+  | 1 => x != y
+  | 2 => BitVec.slt x y
+  | 3 => BitVec.sle x y
+  | 4 => x.toInt > y.toInt
+  | 5 => x.toInt ≥  y.toInt
+  | 6 => BitVec.ult x y
+  | 7 => BitVec.ule x y
+  | 8 => x.toNat > y.toNat
+  | 9 => x.toNat > y.toNat
+  /- ceq/cne = SV case in/equality ("===")
+    tests 4-state logical in/equality (0, 1, x, z)
+    TODO: model the concept.
+  -/
+  | 10 => x == y
+  | 11 => x == y
+  /- weq/wne = SV wild card inequality ("==?")
+    also based on 4-state logical in/equality,
+    with x and z values acting as wild cards.
+    TODO: model the concept.
+  -/
+  | 12 => x == y
+  | 13 => x == y
+  | _ => false
+
+def mods (x y : BitVec w) : BitVec w :=
+  BitVec.smod x y
+
+def modu (x y : BitVec w) : BitVec w :=
+  BitVec.umod x y
+
+def mul {w : Nat} (l : List (BitVec w)) : BitVec w :=
+  List.foldr BitVec.mul (1#w) l
+
+/- according to the docs the input of mux
+  can be any type. I'll stick to bitvecs for now. -/
+def mux (x y : BitVec w) (cond : Bool) : BitVec w :=
+  match cond with
+  | true => x
+  | false => y
+
+def or {w : Nat} (l : List (BitVec w)) : BitVec w :=
+  List.foldr BitVec.or (BitVec.zero w) l
+
+def parity (x : BitVec w) : Bool :=
+  (BitVec.umod x 2#w) == 1
+
+def replicate (x : BitVec w) (n : Nat) : BitVec (w * n) :=
+  BitVec.replicate n x
+
+def shl (x y : BitVec w) : BitVec w :=
+  x <<< y
+
+def shrs (x y : BitVec w) : BitVec w :=
+  BitVec.sshiftRight' x y
+
+def shru (x y : BitVec w) : BitVec w :=
+  BitVec.ushiftRight x y.toNat
+
+def sub (x y : BitVec w) : BitVec w :=
+  x - y
+
+def xor {w : Nat} (l : List (BitVec w)) : BitVec w :=
+  List.foldr BitVec.xor (BitVec.zero w) l
+
+end Comb
+
+namespace MLIR2Comb
+
+section Dialect
+
+inductive Op
+| add (w : Nat) -- n is the number of arguments
+| and (w : Nat)
```

Comment:
Is this right? This seems to be a simple binary `add`, rather than the variadic one